aptos-move/framework/aptos-stdlib/sources/cryptography/groups.move

```diff
@@ -0,0 +1,744 @@
+    ///
+    /// Function `serialize_element_compressed<BLS12_381_G2>` and `deserialize_element_compressed<BLS12_381_G2>`
+    /// assumes a 96-byte encoding `[b_0, ..., b_95]` of an `Element<BLS12_381_G2>` with the following rules.
+    /// - `b_95 & 0x40` is the infinity flag.
+    /// - The infinity flag is 1 if and only if the element is the point at infinity.
+    /// - The infinity flag is 0 if and only if the element is a point `(x,y)` on curve `E(Fq2)`, with the following rules.
+    ///     - `[b_0, ..., b_95 & 0x3f]` is a 96-byte little-endian encoding of `x=(x_0+x_1*u)`.
+    ///         - `[b_0, ..., b_47]` is a 48-byte little-endian encoding of `x_0`.
+    ///         - `[b_48, ..., b_95 & 0x3f]` is a 48-byte little-endian encoding of `x_1`.
+    ///     - `b_95 & 0x80` is the positiveness flag.
+    ///     - The positiveness flag is 1 if and only if `y > -y`.
+    ///         - Here `a=(a_0+a_1*u)` is considered greater than `b=(b_0+b_1*u)` if `a_1>b_1 OR (a_1=b_1 AND a_0>b_0)`.
+    struct BLS12_381_G2 {}
+
+    /// `BLS12_381_Gt` represents a group used in BLS12-381 pairing.
+    /// The group is a prime-order group on an `Fq12`.
+    /// `Fq12` is an extension field of `Fq6`, constructed as `Fq12=Fq6[w]/(w^2-v)`.
+    /// `Fq6` is an extension field of `Fq2`, constructed as `Fq6=Fq2[v]/(v^2-u-1)`.
+    /// `Fq2` is an extension field of `Fq`, constructed as `Fq2=Fq[u]/(u^2+1)`.
+    /// `Fq` is a finite field with `q=0x1a0111ea397fe69a4b1ba7b6434bacd764774b84f38512bf6730d2a0f6b0f6241eabfffeb153ffffb9feffffffffaaab`.
+    /// The order of the group `r` is 0x73eda753299d7d483339d80809a1d80553bda402fffe5bfeffffffff00000001.
+    /// There exists a bilinear mapping from `(BLS12_381_G1, BLS12_381_G2)` to `BLS12_381_Gt`.
+    ///
+    /// A `Scalar<BLS12_381_G2>` is an integer between 0 and `r-1`.
+    ///
+    /// Function `deserialize_scalar<BLS12_381_Gt>` and `serialize_scalar<BLS12_381_Gt>`
+    /// assumes a 32-byte little-endian encoding of a `Scalar<BLS12_381_Gt>`.
+    ///
+    /// An `Element<BLS12_381_Gt>` is an element in `Gt`.
+    ///
+    /// Function `serialize_element_uncompressed<BLS12_381_Gt>` and `deserialize_element_uncompressed<BLS12_381_Gt>`,
+    /// as well as `serialize_element_ompressed<BLS12_381_Gt>` and `deserialize_element_compressed<BLS12_381_Gt>`,
+    /// assume a 576-byte encoding `[b_0, ..., b_575]` of an `Element<BLS12_381_Gt>`, with the following rules.
+    ///     - Assume the given element is `e=c_0+c_1*w` where `c_i=c_i0+c_i1*v+c_i2*v^2 for i=0..1` and `c_ij=c_ij0+c_ij1*u for j=0..2`.
+    ///     - `[b_0, ..., b_575]` is a concatenation of 12 encoded `Fq` elements: `c_000, c_001, c_010, c_011, c_020, c_021, c_100, c_101, c_110, c_111, c_120, c_121`.
+    ///     - Every `c_ijk` uses a 48-byte little-endian encoding.
+    struct BLS12_381_Gt {}
+
+    /// This struct represents an integer between 0 and `r-1`, where `r` is the order of group `G`.
+    struct Scalar<phantom G> has copy, drop {
```

Comment:
Interesting... I now realize this design has an annoyance... even though scalars for both G1, G2 and GT are the same (because the groups have the same order), this module forces you to use different types....

This means developers will have to awkwardly & unnecessarily convert between `Scalar<G1>` and `Scalar<G2>` types even though they are the same type underneath...

Reply:
I think the best way to avoid this is to have a `BLS12381_Fr` definition for the field of scalars $\mathbb{F}_r$ associated with the BLS12-381 curves.

What do you think?

Reply:
right, i had to do this awkward conversion in a test case...
luckily scalar serialization/deserialization is cheap.

Reply:
Well, even so. It's still a little awkward / developer unfriendly. Let's fix it?

Reply:
It means we are introducing `fields.move` in addition to `groups.move`?

Reply:
Hm, I would leave it inside `groups.move`. It would be easier for developers to find everything there.

Reply:
...but it looks like you opted in for making `Scalar<G1>` swappable with `Scalar<G2>` and `Scalar<GT>` which also seems to work, no?

Reply:
we opt to use `BLS12_381_Fr`!